src/helpers/messaging.rs

```diff
@@ -80,10 +80,16 @@ impl<F: Field> Message for F {
 pub struct Gateway {
     /// Sender end of the channel to send requests to receive messages from peers.
     tx: mpsc::Sender<ReceiveRequest>,
-    envelope_tx: mpsc::Sender<(ChannelId, MessageEnvelope)>,
+    envelope_tx: mpsc::Sender<SendRequest>,
     control_handle: JoinHandle<()>,
 }
 
+pub(super) type SendRequest = (
+    ChannelId,
+    MessageEnvelope,
+    oneshot::Sender<Option<PushError>>,
```

Comment:
So this is the real cost of this change.  I'm not sure about this part.

Given that this exists primarily to deal with coding issues, can we limit this to debug builds?

For non-debug builds, a panic on the task that receives messages would be fine.  This is not an error we should be tolerating.